genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GrpcAgentHeartBeatServiceImpl.java

```diff
@@ -0,0 +1,178 @@
+import org.springframework.stereotype.Component;
+import org.springframework.validation.annotation.Validated;
+
+import javax.validation.constraints.NotBlank;
+import java.time.Instant;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * gRPC implementation of AgentHeartBeatService.
+ * Sends heartbeats to the server.
+ * Transparently handles disconnections and stream errors by establishing a new stream.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+@Validated
+class GrpcAgentHeartBeatServiceImpl implements AgentHeartBeatService {
+
+    private static final long HEART_BEAT_PERIOD_MILLIS = 2_000; //TODO make configurable
+    private static final int STREAM_RESET_DELAY_MILLIS = 1000; //TODO make configurable
+
+    private final HeartBeatServiceGrpc.HeartBeatServiceStub client;
+    private final ThreadPoolTaskScheduler taskScheduler;
+
+    private boolean isConnected;
+    private StreamObserver<AgentHeartBeat> requestObserver;
+    private ScheduledFuture<?> heartbeatFuture;
+    private String claimedJobId;
+    private AgentHeartBeat heartBeatMessage;
+
+    GrpcAgentHeartBeatServiceImpl(
+        final HeartBeatServiceGrpc.HeartBeatServiceStub client
+    ) {
+        this.client = client;
+        this.taskScheduler = new ThreadPoolTaskScheduler();
+        this.taskScheduler.setPoolSize(1);
+        this.taskScheduler.initialize();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void start(@NotBlank final String jobId) {
+        if (StringUtils.isNotBlank(this.claimedJobId)) {
+            throw new IllegalStateException("Previously started with a different job id");
+        }
+
+        this.claimedJobId = jobId;
+        this.heartBeatMessage = AgentHeartBeat.newBuilder()
+            .setClaimedJobId(claimedJobId)
+            .build();
+
+        this.heartbeatFuture = taskScheduler.scheduleAtFixedRate(
+            this::sendHeartBeatTask,
+            HEART_BEAT_PERIOD_MILLIS
+        );
+
+        this.requestObserver = client.heartbeat(new ResponseObserver(this));
+
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void stop() {
+        if (this.heartbeatFuture != null) {
+            this.heartbeatFuture.cancel(false);
+            this.heartbeatFuture = null;
+        }
+
+        if (this.requestObserver != null) {
+            this.requestObserver.onCompleted();
+            this.requestObserver = null;
+        }
+
+        this.taskScheduler.shutdown();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized boolean isConnected() {
+        return isConnected;
+    }
+
+    private synchronized void setConnected() {
+        this.isConnected = true;
+    }
+
+    private synchronized void setDisconnected() {
+        this.isConnected = false;
+
+        // Schedule a stream reset
+        this.taskScheduler.schedule(
```

Comment:
Ah! I think you're right this is a bug.

Reply:
Nevermind it should work as is.
A new stream is created. It will either work or throw a stream error, and in the latter case another reset is scheduled.

Reply:
Yes, see `Reset stream after error` test in `GrpcAgentHeartBeatServiceImplSpec`